Fetch charts with GET to benefit from browser cache and conditional requests

betodealmeida opened the pull request:
This is a small PR that does a lot. It changes the initial request for charts (in explore or dashboards) to be done through a `GET` request, greatly improving the loading speed of dashboards. It also moves the caching to the HTTP layer, allowing us to benefit from `Expires` and `ETag` headers for conditional requests.

# The problem

This diagram compares the current flow ("before") with the one implemented by this PR ("after"):

<img width="512" alt="Cache" src="https://user-images.githubusercontent.com/1534870/54342019-1543dc00-4644-11e9-90ea-76862a31e505.png">

## Before

Let's assume Superset is configured with a **1 hour cache**, and also that the data changes on a longer period (daily, eg):

1. User "A" requests a chart from Superset doing a `POST` request with the payload.
2. Superset computes the query and sends it to the DB.
3. DB returns a dataframe.
4. Superset caches the dataframe.
5. Superset serializes the payload and sends it back to user "A".
6. User "A" refreshes the dashboard.
7. Superset finds the dataframe cached.
8. Superset serializes the payload and send it back to user "A".
9. Superset cache expires after 1 hour.
10. User "A" refreshes the dashboard.
11. Superset computes the query and sends it to the DB.
12. DB returns the exact same dataframe.
13. Superset caches the dataframe again.
14. Superset serializes the payload and sends it back to user "A".

There are a few inefficiencies here:

- The browser cache is never used, because it's doing `POST` requests.
- Superset needs to serialize the payload even on a cache hit.
- Data is transferred to the browser even if it hasn't changed.

## After

1. User "A" requests a chart from Superset **doing a `GET` request with the chart id**.
2. Superset computes the query and sends it to the DB.
3. DB returns a dataframe.
4. Superset serializes the dataframe and **caches the HTTP response**.
5. Superset sends the payload to user "A", with an `Expires` header of 1 hour, and an `ETag` header which is a hash of the payload.
6. The browser stores the response in its native cache, and `SupersetClient` caches it also in the [Cache interface](https://developer.mozilla.org/en-US/docs/Web/API/Cache).
7. The user refreshes the dashboard.
8. Because of the `Expires` header and the use of `GET` the **data is read directly from the native browser cache**.
9. Superset cache expires after 1 hour.
10. User "A" refreshes the dashboard. The **native cache is not used, since `Expires` is now in the past**. `SupersetClient` looks for a cached response in the Cache interface, and if one is found, extracts its `ETag`.
11. The **browser requests the chart with an `If-None-Match` header**, containing the hash of the cached response (its `ETag`).
12. Superset computes the query and sends it to the DB.
13. DB returns the exact same dataframe.
14. Superset serializes the dataframe and **caches the HTTP response**.
15. Superset sees that the `ETag` matches the `If-None-Match` header, returning a `304 Not Modified` response.
16. Browser fetches the cached response from the Cache interface.
17. Browser uses the response.

# Notes

- The `GET` request is done only the first time the chart is mounted. Forcing refresh on dashboards and clicking "Run Query" in the Explore views perform `POST` requests, which bypass the cache, and cache the new response. I tested the Explore view and dashboards with filters, and all further interactions are done with `POST`s.

- Since we're caching the HTTP response, we need to verify that the user has permission to read the cached response. This is done by passing a `check_perms` function to the decorator that caches the responses.

- The [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) has no support for conditional responses with ETags. We need to add explicit support in `SupersetClient`. I have a separate PR for that (see https://github.com/apache-superset/superset-ui/pull/119).

- **There is one small downside to this approach**. During the time while `Expires` is still valid, the browser will not perform any requests for cached charts unless the user explicitly refreshes a dashboard or click "Run Query" in the Explore view. If the data is bad, they will see bad data until it expires or they purposefully refresh the chart. In the current workflow, in theory we can purge the cache in this case, since it lives only on the server-side. This is a hypothetical scenario, and we could workaround it by sending a notification to dashboards that one or more charts have bad data and should be refreshed.

john-bodley committed:
[csv-upload] Fixing message encoding (#6971)


(cherry picked from commit 48431ab5b9375a94c5262a0336d9c69e5f01a3ac)

conglei committed:
Enhancement of query context and object. (#6962)

* added more functionalities for query context and object.

* fixed cache logic

* added default value for groupby

* updated comments and removed print

(cherry picked from commit d5b9795f87f79fa2c41e144ffc00fd9586be7657)

khtruong committed:
Sparkline dates aren't formatting in Time Series Table (#6976)

* Exclude venv for python linter to ignore

* Fix NaN error

khtruong committed:
Sparkline dates aren't formatting in Time Series Table (#6976)

* Exclude venv for python linter to ignore

* Fix NaN error

john-bodley committed:
[sql-parse] Fixing LIMIT exceptions (#6963)


(cherry picked from commit 3e076cb60b385e675ed1c9a8053493375e43370b)

betodealmeida committed:
Add decorator for etag cache

betodealmeida committed:
Add decorator for etag cache

michellethomas committed:
Adding custom control overrides (#6956)

* Adding extraOverrides to line chart

* Updating extraOverrides to fit with more cases

* Moving extraOverrides to index.js

* Removing webpack-merge in package.json

* Fixing metrics control clearing metric

(cherry picked from commit e6194051f486e42922dc4e34a861f4490c1062fc)

Commit:
[fix] /superset/slice/id url is too long (#6989)


(cherry picked from commit 6a4d507ab607b01ed324cb3341b71c6fb2cb5c97)

betodealmeida committed:
Fetch charts with GET

betodealmeida committed:
Fetch charts with GET

betodealmeida committed:
Small fixes

betodealmeida committed:
Small fixes

betodealmeida committed:
Fix typo

betodealmeida committed:
Fix typo

betodealmeida committed:
Compute correct cache key; fix logging

betodealmeida committed:
Compute correct cache key; fix logging

betodealmeida committed:
Check perms on cached response

betodealmeida committed:
Check perms on cached response

michellethomas committed:
Changing time table viz to pass formatTime a date (#7020)

(cherry picked from commit 7f3c145b1f5a4e2d8b95982119503e98772e2c47)

michellethomas committed:
Changing time table viz to pass formatTime a date (#7020)

(cherry picked from commit 7f3c145b1f5a4e2d8b95982119503e98772e2c47)

xtinec committed:
Fix the white background shown in SQL editor on drag (#7021)

This PR sets the background-color css property on `.ace_scroller` instead of `.ace_content` to prevent the white background shown during resizing of the SQL editor before drag ends.

john-bodley committed:
[db-engine-spec] Aligning Hive/Presto partition logic (#7007)


(cherry picked from commit 05be86611785fef2904992e4e7d31dce23f1c51b)

betodealmeida committed:
Show tooltip with time frame (#6979)

betodealmeida committed:
Fix time filter control (#6978)

betodealmeida committed:
Revert change

betodealmeida committed:
Revert change

betodealmeida committed:
If perms fail, return naked response

betodealmeida committed:
If perms fail, return naked response

betodealmeida commented:
:eyes @diggidydave

betodealmeida committed:
Fix lint

betodealmeida committed:
Fix lint

graceguo-supercat commented:
you sure GET request can handle `this.props.formData`? in airbnb many, many chart's formData are longer than 4k chars :)

betodealmeida commented:
@graceguo-supercat:

> you sure GET request can handle this.props.formData? in airbnb many, many chart's formData are longer than 4k chars :)

Thanks, I'm aware of that problem. Here the `GET` request has only the chart ID, and the form data is read from the saved chart (`params` in the `slices` table is used).

> why not use Redis cache query results, in aribnb the round trip to fetch data from Redis is only 600ms~800ms.

The decorator is using Redis for the server-side caching, but it's caching the HTTP response instead of the dataframe (saving time that's spent in serialization). But I'm also using the native browser cache (through the `Expires` header) and the new "Cache interface API" (in a separate PR that I'm finishing).

> i think we should use etag for dashboard metadata (like dashboard layout, huge json blob...)

Everywhere! :)

graceguo-supercat commented:
`Here the GET request has only the chart ID, and the form data is read from the saved chart (params in the slices table is used).`
how about in dashboard with filter, each chart's query is not the saved params.
Also, you still have a `formData` in request parameters right? this is the blocker that cause issue.

betodealmeida commented:
> how about in dashboard with filter, each chart's query is not the saved params? the actual query is formData overwrite saved chart's params.

Ah, you're right. The filter box works when changed, since it does a `POST`. But in the initial load it's not taken into consideration. Let me see how I can fix that.

> Also, you still have a formData in GET request parameters right (you can see request parameters from browser location bar)? this is the blocker that cause issue.

Yes, but it has only the `slice_id` in it right now, eg: `form_data: {"slice_id":78}`. I'll see if I can append any additional parameters that are set by filters, this should keep it small.

betodealmeida commented:
@graceguo-supercat I tested the interaction with filter boxes and it's not working. I'll work on fixing it.

thunter009 committed:
[WIP] fix user specified JSON metadata not updating dashboard on refresh (#7027)


(cherry picked from commit cc58f0e661044e95c7c86d0da8d77a0a6640efe7)

williaster commented:
@betodealmeida this seems pretty complicated on top of data requests that are already complicated 
 and error prone 🙉 

My first question to gauge whether it's worthwhile is: what are the speedup times you are seeing for existing approach vs your new approach? (ideally for multiple dashboards of varying size)

betodealmeida commented:
@williaster the speedup will greatly depend on how often the data changes, how big the payload is (bignum vs deck.gl varies significantly), the duration of the cache, and how slow the network is. I can (and have) run tests against the example dashboards, but I don't think they would be significative, since they don't cover all the real life use cases.

I think the question we should ask here is: "given that this is clearly an improvement, how can we make it bug free?"

conglei committed:
fix inaccurate data calculation with adata rolling and contribution (#7035)


(cherry picked from commit 0782e831cd37f665a2838119d87c433269f1b36b)

john-bodley commented:
@betodealmeida just to clarify step 4 (and per your diagram) you have: 

- **Before**: Superset caches the dataframe.
- **After**: Superset serializes the dataframe and caches the **HTTP response**.

yet in the code it still seems like we're caching the result set from the database and thus I wonder if the diagram and thus **After** phase should mention that there's an additional entry in the server cache, i.e., after step (4) the server-side cache would contain:
- The cached result set (as a dataframe).
- The cached `superset/explore_json` HTTP response which includes server-side Python visualization specific mutations.

Note I'm not saying this is wrong as I strongly believe that the database response should be cache given that represents the bulk of the compute, I just wanted to get clarity on the logic.

williaster commented:
@betodealmeida this is a big change that has the potential to impact many many users. If you're unable to provide numbers that indicate that this is strictly an improvement (or at a minimum no regressions), I'm a bit reluctant to introduce this additional complexity. 

I think it's part of the expected work of a feature like this to demonstrate the effects with real-life examples. It seems like you should be able to use real Lyft dashboards, dashboards from the "example datasets", etc., and you can throttle network speed using dev tools. 

Another concern I have is the impact on the # of requests. We've needed to introduce domain sharding because of the large number of simultaneous requests made by larger dashboards, and this potential **DOUBLES** that number, so I would want to see perf numbers that demonstrate no regressions for that case as well.

mistercrunch commented:
Not directly related, but on the topic of optimization around caching, it would be an extra win to make the caching call `async`. Maybe a first use case for the async/await syntax in Superset.

williaster commented:
@mistercrunch `fetch` (`SupersetClient`) is `async`? do you just mean the `async/await` syntax?

mistercrunch commented:
@williaster I'm thinking about something else: what I mean on the server side, in python, making the `cache.set(...)` call async, meaning while a thread is pushing to the caching backend, the web server can start streaming the response to the client at the same time. There's no need to wait on the caching to start sending the response back...

betodealmeida committed:
Compute cache key from all form data

betodealmeida committed:
Compute cache key from all form data

betodealmeida commented:
@williaster:

> @betodealmeida this is a big change that has the potential to impact many many users. If you're unable to provide numbers that indicate that this is strictly an improvement (or at a minimum no regressions), I'm a bit reluctant to introduce this additional complexity.

> I think it's part of the expected work of a feature like this to demonstrate the effects with real-life examples. It seems like you should be able to use real Lyft dashboards, dashboards from the "example datasets", etc., and you can throttle network speed using dev tools.

Sure, I will give the numbers of the example dashboards and some of the Lyft dashboards. My point was that, considering that this is a strict improvement, defining a threshold to accept the changes seems arbitrary to me.

Of course I agree that if this causes a regression or no significant improvement we should not do it. Maybe it's not clear, but with this PR the client will **always** issue a smaller number of requests, and a percentage of those requests will receive body-less responses. Combined with the fact that we move the server cache closer to the user, there shouldn't be any regressions with this PR, unless I'm doing something stupid (which I've done in the past).

Also, keep in mind that this PR is against the `lyftga` branch, so we expect to test it before merging into `master`.

> Another concern I have is the impact on the # of requests. We've needed to introduce domain sharding because of the large number of simultaneous requests made by larger dashboards, and this potential DOUBLES that number, so I would want to see perf numbers that demonstrate no regressions for that case as well.

Sorry, why do you think this would double the number of requests? The number of requests should be strictly equal or smaller: resources within the lifetime of the cache are no longer requested because the browser reads directly from its cache, and conditional requests are still a single request. The server will either return a normal response (`200 OK`) or a `304 Not Modified` without body. There's no pre-flight request.

betodealmeida committed:
Pass extra_filters in GET request

betodealmeida committed:
Pass extra_filters in GET request

betodealmeida commented:
@john-bodley not sure if I understand your question. Currently we cache the dataframe in `viz.BaseViz.get_df_payload`. I haven't touched that code, but I added an additional cache storing the `Response` object instead. This has a few benefits:

- It skips unpickling the dataframe and serializing it to json.
- We don't need to recompute the ETag every time we read from the cache, since it's stored in the cached response.

Eventually we can remove the dataframe caching, but I'd rather do that in a separate PR.

betodealmeida commented:
@graceguo-supercat I changed the code so that in the `GET` request we pass:

1. the `slice_id`
2. any `extra_filters`

This way the `GET` request is still relatively small (and we can switch to [Rison](https://github.com/apache/incubator-superset/pull/5152) for a smaller URL if needed), and the caching mechanism works with dashboards that have filters saved.

betodealmeida committed:
Fix pylint

betodealmeida committed:
Fix pylint

betodealmeida committed:
Fix flake8

betodealmeida committed:
Fix flake8

john-bodley committed:
[sqlparse] Fixing table name extraction for ill-defined query (#7029)


(cherry picked from commit 07c340cf8203f13222f16efad1e55e202deb1865)

john-bodley commented:
@betodealmeida my point was that I felt that the picture and description doesn't accurately reflect what your code is actually doing, i.e., that there's actually two server-side cache keys associated with each chart. They're both using the same underlying cache but the diagram and steps don't accurately reflect this. 

I agree with your logic but thought maybe the steps should be more explicit, e.g., 

**After**

3. DB returns a dataframe.
4. Superset caches the dataframe.
5. Superset serializes the payload.
6. Superset caches the the HTTP response associated with the payload.
7. Superset sends the payload to user "A", with an Expires header of 1 hour, and an ETag header which is a hash of the payload.

Commit:
[fix] Cursor jumping when editing chart and dashboard titles (#7038)


(cherry picked from commit fc1770f7b79a4d8815b646b46390fabf190c3815)

betodealmeida commented:
@john-bodley you're right. I ended up describing in "after" the workflow once we remove the dataframe cache.

Commit:
[fix] explore chart from dashboard missed slice title (#7046)


(cherry picked from commit a6d48d4052839286aec725d51303b3b2bf6e8dd4)

DiggidyDave commented on superset/utils/decorators.py:
```diff
@@ -29,3 +35,51 @@ def stats_timing(stats_key, stats_logger):
+                cache_key = wrapper.make_cache_key(f, *key_args, **key_kwargs)
+                response = cache.get(cache_key)
+            except Exception:  # pylint: disable=broad-except
+                logging.exception('Exception possibly due to cache backend.')
```
any chance you can tighten up this exception handling and messaging? Do we really need to disable the linter here? If we handle distinct exception types could we get logging that is more definitive than "possibly due to...."? Sorry to be a stickler, but caching issues can be super frustrating, and this sort of thing can go a long way to short-circuiting debugging work.

john-bodley commented on superset/utils/decorators.py:
```diff
@@ -29,3 +35,51 @@ def stats_timing(stats_key, stats_logger):
+                key_kwargs.update(request.args)
+                cache_key = wrapper.make_cache_key(f, *key_args, **key_kwargs)
+                response = cache.get(cache_key)
+            except Exception:  # pylint: disable=broad-except
```
Maybe there is merit in following the pattern in Flask-Caching per [here](https://github.com/sh4nks/flask-caching/blob/master/flask_caching/__init__.py#L403) where the error is re-raised in debug mode.

betodealmeida replied:
Ah, good point. I'll do that.

DiggidyDave replied:
Can we remove the pylint exception while we're in here? I'm not seeing that anywhere else in the codebase despite similarly structured except blocks.

john-bodley replied:
@DiggidyDave the reason you're probably not seeing it is we have [this](https://github.com/apache/incubator-superset/blob/master/superset/views/base.py#L17) defined in a number of files. It's a place holder when we enforced that new code could not have any `pylint` warnings.

DiggidyDave replied:
Gotcha, thanks for the info. If this is the first new code that isn't grandfathered into a whole file disable directive, should we consider either 1) making this adhere to the linter rules (ie don't disable it here) if this is a warning we care about or 2) adding a blanket exclusion to `.pylintrc` if this is a warning we never care about (which seems to be the case in practice)?

If this is something that we globally don't care about at this particular moment, it seems like a good idea to just codify and enforce that in one place in the central config.

john-bodley replied:
@DiggidyDave catching a broad exception is generally bad hence why the `pylint` rule exists, i.e.,
```
try:
     int('x')
except ValueError:
    ...
```

is preferred over,
```
try:
     int('x')
except Exception:
    ...
```
That said there are rare occasions when using the later form make sense, i.e., i) the caller may be agnostic of the underlying function being called, or ii) all exceptions need to correctly corralled. In this example I think the current logic is correct. Additionally it mimics [Flask-Caching](https://github.com/sh4nks/flask-caching/blob/master/flask_caching/__init__.py#L387).

DiggidyDave replied:
I understand why they are bad, and I agree this is ok here. However, linter rules that are _always_ ignored in an ad-hoc way (ie not configured globally) sometimes can raise a red flag wrt code hygiene for me, so I thought I'd bring it up here while I'm ramping up on the codebase. :-) Thanks for your responses! I think we're on the same page.

betodealmeida committed:
Use ETags even if no cache is set

betodealmeida committed:
Use ETags even if no cache is set

betodealmeida committed:
Handle adhoc filters

betodealmeida committed:
Handle adhoc filters

john-bodley committed:
[datasource] Ensuring consistent behavior of datasource editing/saving. (#7037)

* Update datasource.py

* Update datasource.py

(cherry picked from commit c771625f1068d3a7f41e6bced14b0cbdbf9962cc)

michellethomas committed:
Adding warning message for sqllab save query (#7028)


(cherry picked from commit ead3d48133e7e1ab8b91d51e561544a544b4eaad)

williaster commented on superset/assets/src/chart/chartAction.js:
```diff
@@ -246,6 +247,16 @@ export function runQuery(formData, force = false, timeout = 60, key) {
 }
 
+export const FETCH_CHART = 'FETCH_CHART';
+export function fetchChart(formData, force = false, timeout = 60, key) {
```
can you give these more specific names, and possibly leave a comment as to how they are different? 

If I was a new dev to Superset, I would be very confused why there were two nearly identical methods that only change GET/POST. e.g., `fetchCachedChart` (or similar) and `runChartQuery` might be more clear.

betodealmeida replied:
Will do.

Commit:
handle null column_name in sqla and druid models

(cherry picked from commit 2ff721ae072b8d69c5cabddc3e1a388a596b1b6f)

khtruong committed:
feat: add ability to change font size in big number (#7003)

* Add ability to change font sizes in Big Number

* rename big number to header

* Add comment to clarify font size values

betodealmeida committed:
Allow LIMIT to be specified in parameters (#7052)

john-bodley committed:
[missing values] Removing replacing missing values (#4905)


(cherry picked from commit 61add606ca16a6ba981ccde864b121f5464b697a)

Commit:
[forms] Fix handling of NULLs

(cherry picked from commit e83a07d3dfda350cc44041cb6cbaec4510887902)

mistercrunch committed:
Fix filter_box migration PR #6523 (#7066)

* Fix filter_box migration PR #6523

* Fix druid-related bug

(cherry picked from commit b210742ad24d01ca05bc58ca3342c90e301fe073)

enricoberti committed:
[SQL Lab] Improved query and results tabs rendering reliability (#7082)

closes #7080

(cherry picked from commit 9b58e9f4920ef424e5b545dcbb4726e22bed5982)

betodealmeida committed:
Raise in debug mode

betodealmeida committed:
Raise in debug mode

betodealmeida committed:
Rename actions

betodealmeida committed:
Rename actions

betodealmeida committed:
Fix integration tests

betodealmeida committed:
Fix integration tests

betodealmeida committed:
Do POST request on new charts

betodealmeida committed:
Do POST request on new charts

betodealmeida committed:
Set extra/adhoc filters only in GET requests

betodealmeida committed:
Set extra/adhoc filters only in GET requests

betodealmeida committed:
Raise if check_perms fails

betodealmeida committed:
Raise if check_perms fails

betodealmeida committed:
Refactor auth

betodealmeida committed:
Refactor auth

betodealmeida committed:
Fix flake8

betodealmeida committed:
Fix flake8

michellethomas committed:
Use metric name instead of metric in filter box (#7106)


(cherry picked from commit 003364e74ea70cad1a4a6e784933fe8bef4c78ec)

betodealmeida committed:
Fix js unit tests

betodealmeida committed:
Fix js unit tests

betodealmeida committed:
Fix js unit tests that fail in lyftga

betodealmeida committed:
Fix js unit tests that fail in lyftga

betodealmeida committed:
Fix js

betodealmeida committed:
Fix js

xtinec committed:
SQL editor layout makeover (#7102)

This PR includes the following layout and css tweaks:
- Using flex to layout the north and south sub panes of query pane so resizing works properly in both Chrome and Firefox
- Removal of necessary wrapper divs and tweaking of css in sql lab so we can scroll to the bottom of both the table list and the results pane
- Make sql lab's content not overflow vertically and layout the query result area to eliminate double scroll bars
- css tweaks on the basic.html page so the loading animation appears in the center of the page across the board

(cherry picked from commit 71f1bbd2ec59b99d6ba6d9a4a2f9cfceaf922b80)

xtinec committed:
SQL editor layout makeover (#7102)

This PR includes the following layout and css tweaks:
- Using flex to layout the north and south sub panes of query pane so resizing works properly in both Chrome and Firefox
- Removal of necessary wrapper divs and tweaking of css in sql lab so we can scroll to the bottom of both the table list and the results pane
- Make sql lab's content not overflow vertically and layout the query result area to eliminate double scroll bars
- css tweaks on the basic.html page so the loading animation appears in the center of the page across the board

(cherry picked from commit 71f1bbd2ec59b99d6ba6d9a4a2f9cfceaf922b80)

mistercrunch committed:
Bump python lib croniter to an existing version (#7132)

Package maintainers should really never delete packages, but it appears
this happened with croniter and resulted in breaking our builds.

This PR bumps to a more recent existing version of the library

(cherry picked from commit 215ed392a11598eac228f57341dbfd232cf770e3)

betodealmeida commented:
@xtinec can we merge this into `lyftga`? It will help test the cache warmup.

betodealmeida commented:
@williaster @john-bodley is it ok to merge this? I've addressed all the issues.

betodealmeida committed:
Revert PR #6933 (#7162)

john-bodley commented:
@betodealmeida I was wondering whether there was any qualitative or quantitative evidence of any performance improvements?

john-bodley commented on superset/utils/decorators.py:
```diff
@@ -29,3 +35,60 @@ def stats_timing(stats_key, stats_logger):
+                response.expires = response.last_modified + timedelta(seconds=max_age)
+                response.add_etag()
+                try:
+                    cache.set(cache_key, response, timeout=max_age)
```
@betodealmeida note in [Flask-Caching](https://github.com/sh4nks/flask-caching/blob/5599d7aafdfa36f1f3cfabfd19a13942c98352d4/flask_caching/backends/simple.py#L19), 

> A timeout of 0 indicates that the cache never expires.

which is inconsistent with the result expiry if the `max_age` is zero, i.e., the backend cache never expires whereas the response cache immediately expires.

betodealmeida replied:
@john-bodley good point. I'll fix this.

williaster commented:
ditto john's question about something that demonstrates the perf improvements/as I asked earlier. otherwise lgtm 👍

betodealmeida commented:
@john-bodley @williaster 

Sorry, I ran some numbers and forgot to post them here. I did a few tests with 4 examples dashboards:

![Improvement over current approach](https://user-images.githubusercontent.com/1534870/55351397-157d0d80-5473-11e9-8cd7-3004cf927c7d.png)

For the "Misc Charts", eg, the first run is more than 20% faster, and the second and third requests are more than 50% faster.

Here are the numbers:

https://docs.google.com/spreadsheets/d/1bliQs6fnqni0lcNuxKRQ-A22RhOTiOI6OHZaR2tClh0/edit?usp=sharing

betodealmeida committed:
Merge

betodealmeida committed:
Fix bad merge

betodealmeida committed:
Use far future when max_age=0

williaster commented:
thanks for the benchmarks! 

will let @john-bodley sign off since [he had the last requested change](https://github.com/apache/incubator-superset/pull/7032#discussion_r270540384).

betodealmeida committed:
Merge

betodealmeida committed:
Merge branch 'lyft-VIZ-334' into VIZ-334

betodealmeida committed:
Fix conflict

betodealmeida committed:
Merge branch 'lyft-VIZ-334' into VIZ-334

mistercrunch commented:
I was working on this tricky bug and ended up here as the cause for it. The issue is around the fact that the `formData` may need to get sanitized by frontend-related logic.

The bug or symptom is the "Genders by State" example started showing an extra (3rd) metric (`sum__num`) on top of `Boys` and `Girls`.

There's a lot going on related to this, but a key point here is that we have control-related logic like `default` and things around control panel that sanitizes `formData`. In that particular case, the example's form data is malformed with both `metrics` and `metric`. In the explore view, or in a normal POST request, the formData gets sanitized:
* missing keys get filled with the control's default
* extra keys get deleted
* more stuff

Now another thing that compounds here is that `viz.py` blindly looks at all METRIC_KEYS to help craft the query_obj. So the combo of these things create the issue:
* example is malformed, it should have only `metrics` and NO `metric` key (easy to fix)
* in GET mode, the formData does not get "sanitized" by the frontend
* viz.py could be more prescriptive about the metrics it builds into the query object

Now it's pretty clear that addressing any of these 3 things would fix my symptom, but point 2 on its own is worrisome. It can lead to intricate issues over time. Say I save a chart, and after that I add a new control to that vizType, in the context of the `GET`, it won't get the right default.

Good news is I'm working on a refactor https://github.com/apache/incubator-superset/pull/7350 to help and clean up all of the control / formData processing logic. It grew out of control over time. Now the assumption is that all request would make it through this logic, and I'm realizing that's not the case, at least since this PR.

Ideas?

mistercrunch commented on superset/utils/decorators.py:
```diff
@@ -29,3 +41,61 @@ def stats_timing(stats_key, stats_logger):
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            # check if the user can access the resource
+            check_perms(*args, **kwargs)
```
I was digging around to try and figure out where the datasource access permission is done nowadays, and found it here in the etag_cache decorator. I feel like it's not the right place for it.

I understand that this needs to happen prior to reading from cache, but maybe it should be done as a prior decorator, or maybe both of these routines should be done inside a method instead of decorators, to avoid calling get_viz twice.

mistercrunch commented on superset/views/core.py:
```diff
@@ -100,6 +101,46 @@ def is_owner(obj, user):
+    form_data, slc = get_form_data(slice_id, use_slice_data=True)
+    datasource_type = slc.datasource.type
+    datasource_id = slc.datasource.id
+    viz_obj = get_viz(
```
I think `get_viz` gets called at least two times now (here and in the view itself)

mistercrunch commented:
Found some other issues here that I wanted to raise ^^^

Also I noticed that the big "merge" on `master` of this and much more stuff got actually done on a single commit instead of a proper merge 538776b47025f8ff97f293f943736cb3633b2702

In the future, `lyftga` branches and the likes should be **merged**, not squashed and merged as we loose tons of history. For instance if we wanted to revert this PR, there's no single commit in master we can address, we'd have to revert all of 538776b47025f8ff97f293f943736cb3633b2702 or get really creative...

DiggidyDave commented:
@mistercrunch agreed, and after that lyftga branch all of our commits were merged individually (from `lyft-release-sp8`.  We have now switched to working on and continuously deploying from `master` so this should not longer even _be able to_ happen. ;-)
